File: FtcRobotController/src/main/java/org/firstinspires/ftc/robotcontroller/moeglobal/slam/SlamUsbHandler.java
package org.firstinspires.ftc.robotcontroller.moeglobal.slam;

import android.content.Context;
import android.content.pm.PackageManager;
import android.hardware.usb.*;
import android.util.Log;

import java.util.*;

import static org.firstinspires.ftc.robotcontroller.moeglobal.ActivityReferenceHolder.activityRef;
import static org.firstinspires.ftc.robotcontroller.moeglobal.slam.Constants.CHUNK_SIZE;
import static org.firstinspires.ftc.robotcontroller.moeglobal.slam.Constants.SMALL_TIMEOUT;

public class SlamUsbHandler {
    public static UsbManager usbManager;

    public static void uploadByteArray(UsbDevice device, byte[] arr) {
        UsbDeviceConnection deviceConnection = getDeviceConnection(device);
        UsbInterface usbInterface = getUsbInterface(device);
        UsbEndpoint outgoingEndpoint = getOutgoingEndpoint(usbInterface);
        sendToEndpoint(deviceConnection, usbInterface, outgoingEndpoint, arr);
    }

    private static void sendToEndpoint(UsbDeviceConnection connection, UsbInterface usbInterface, UsbEndpoint outPoint, byte[] arr) {
        if (!connection.claimInterface(usbInterface, true)) {
            throw new IllegalStateException("Failed to claim Usb Interface");
        }

        List<byte[]> listToBeSplit = divideArray(arr, CHUNK_SIZE);
        for (byte[] bytes : listToBeSplit) {
            connection.bulkTransfer(outPoint, bytes, CHUNK_SIZE, SMALL_TIMEOUT);
        }

        connection.releaseInterface(usbInterface);
        connection.close();
    }

    private static List<byte[]> divideArray(byte[] source, int chunkSize) {

        List res = new ArrayList();
        int start = 0;

        for (int i = 0; i < (int) Math.ceil(source.length / (double) chunkSize); i++) {
            res.add(Arrays.copyOfRange(source, start, start + chunkSize));
            start += chunkSize;
        }

        return res;
    }

    private static UsbInterface getUsbInterface(UsbDevice device) {
        return device.getInterface(0);
    }

    private static UsbEndpoint getOutgoingEndpoint(UsbInterface usbInterface) {
        for (int k = 0; k < usbInterface.getEndpointCount(); k++) {
            UsbEndpoint uep = usbInterface.getEndpoint(k);
            if (uep.getDirection() == UsbConstants.USB_DIR_OUT) {
                return uep;
            }
        }
        throw new IllegalStateException("Outgoing endpoint not found");
    }

    public static UsbDeviceConnection getDeviceConnection(UsbDevice device) {
        UsbDeviceConnection connection = usbManager.openDevice(device);
        if (connection == null) throw new IllegalStateException("Failed to Open Device");
        return connection;
    }

    public static UsbDevice getDevice(int vid) {
        boolean b = activityRef.get().getPackageManager().hasSystemFeature(PackageManager.FEATURE_USB_HOST);
        boolean b1 = activityRef.get().getPackageManager().hasSystemFeature(PackageManager.FEATURE_USB_ACCESSORY);
        UsbAccessory[] accessoryList = usbManager.getAccessoryList();
        HashMap<String, UsbDevice> deviceList = usbManager.getDeviceList();
//        usbManager.openDevice()
//        UsbManager manager = (UsbManager) getSystemService(Context.USB_SERVICE);

//        HashMap<String, UsbDevice> deviceList = manager.getDeviceList();
        Iterator<UsbDevice> deviceIterator = deviceList.values().iterator();
        while (deviceIterator.hasNext()) {
            UsbDevice device = deviceIterator.next();

//            manager.requestPermission(device, mPermissionIntent);
            String Model = device.getDeviceName();

            int DeviceID = device.getDeviceId();
            int Vendor = device.getVendorId();
            int Product = device.getProductId();
            int Class = device.getDeviceClass();
            int Subclass = device.getDeviceSubclass();

        }
        Log.e("devices", String.valueOf(deviceList.size()));
        for (UsbDevice device : deviceList.values()) {
            Log.e("devices", String.valueOf(device.getVendorId()));
        }
        for (UsbDevice device : deviceList.values()) {
            if (device.getVendorId() == vid) {
                return device;
            }
        }
        return null;
    }

    static void init(Context context) {
        usbManager = (UsbManager) context.getSystemService(Context.USB_SERVICE);
    }

}

File: FtcRobotController/src/main/java/org/firstinspires/ftc/robotcontroller/moeglobal/slam/SlamUsbHandler.kt
package org.firstinspires.ftc.robotcontroller.moeglobal.slam

import android.content.Context
import android.content.pm.PackageManager
import android.hardware.usb.*
import android.util.Log
import org.firstinspires.ftc.robotcontroller.moeglobal.ActivityReferenceHolder
import org.firstinspires.ftc.robotcontroller.moeglobal.divideArray

object SlamUsbHandler {
    var usbManager: UsbManager? = null
    fun uploadByteArray(device: UsbDevice, arr: ByteArray) {
        val deviceConnection = getDeviceConnection(device)
        val usbInterface = getUsbInterface(device)
        val outgoingEndpoint = getOutgoingEndpoint(usbInterface)
        sendToEndpoint(deviceConnection, usbInterface, outgoingEndpoint, arr)
    }

    private fun sendToEndpoint(connection: UsbDeviceConnection, usbInterface: UsbInterface, outPoint: UsbEndpoint, arr: ByteArray) {
        check(connection.claimInterface(usbInterface, true)) { "Failed to claim Usb Interface" }
        val listToBeSplit = arr.divideArray(Constants.CHUNK_SIZE)
        for (bytes in listToBeSplit) {
            connection.bulkTransfer(outPoint, bytes, Constants.CHUNK_SIZE, Constants.SMALL_TIMEOUT)
        }
        connection.releaseInterface(usbInterface)
        connection.close()
    }


    private fun getUsbInterface(device: UsbDevice): UsbInterface {
        return device.getInterface(0)
    }

    private fun getOutgoingEndpoint(usbInterface: UsbInterface): UsbEndpoint {
        for (k in 0 until usbInterface.endpointCount) {
            val uep = usbInterface.getEndpoint(k)
            if (uep.direction == UsbConstants.USB_DIR_OUT) {
                return uep
            }
        }
        throw IllegalStateException("Outgoing endpoint not found")
    }

    fun getDeviceConnection(device: UsbDevice?): UsbDeviceConnection {
        return usbManager!!.openDevice(device) ?: throw IllegalStateException("Failed to Open Device")
    }

    fun getDevice(vid: Int): UsbDevice? {
//        val b = ActivityReferenceHolder.activityRef.get()!!.packageManager.hasSystemFeature(PackageManager.FEATURE_USB_HOST)
//        val b1 = ActivityReferenceHolder.activityRef.get()!!.packageManager.hasSystemFeature(PackageManager.FEATURE_USB_ACCESSORY)
//        val accessoryList = usbManager!!.accessoryList
        val deviceList = usbManager!!.deviceList
        //        usbManager.openDevice()
        //        UsbManager manager = (UsbManager) getSystemService(Context.USB_SERVICE);
        //        HashMap<String, UsbDevice> deviceList = manager.getDeviceList();
//        val deviceIterator: Iterator<UsbDevice> = deviceList.values.iterator()
//        while (deviceIterator.hasNext()) {
//            val device = deviceIterator.next()
            //            manager.requestPermission(device, mPermissionIntent);
//            val Model = device.deviceName
//            val DeviceID = device.deviceId
//            val Vendor = device.vendorId
//            val Product = device.productId
//            val Class = device.deviceClass
//            val Subclass = device.deviceSubclass
//        }
        Log.e("devices", deviceList.size.toString())
        for (device in deviceList.values) {
            Log.e("devices", device.vendorId.toString())
        }
        for (device in deviceList.values) {
            if (device.vendorId == vid) {
                return device
            }
        }
        return null
    }

    fun init(context: Context) {
        usbManager = context.getSystemService(Context.USB_SERVICE) as UsbManager
    }
}